File: apps/api/app/services/media.py
"""Disponibilité des médias (contrat parcours §1) — miroir de `hasNativeAudio` / `hasPitchRef` (packages/core).

- `mediaIndex` : chemins relatifs des fichiers présents (`audio/…`, `pitch/…`, `img/…`), côté API calculé à partir
  de CONTENT_DIR (+ médias du studio, pas encore publiés).
- `has_native_audio(concept)` : au moins une piste `source: "native"` dont `src` est dans l'index.
- `has_pitch_ref(concept|step)` : `pitchRef` (étape) ou `pitch` (concept) présent dans l'index.
"""

from collections.abc import Collection, Mapping
from pathlib import Path
from typing import Any

from app.services.content import Pack, scan_media

TONE_STEP_TYPES = frozenset({"tone_identify", "tone_minimal_pair", "tone_produce"})
SPEECH_STEP_TYPES = frozenset({"speak_repeat", "tone_produce"})


class _AllMedia(frozenset[str]):
    """Index « tout est présent » (`mediaIndex` null côté core : fichiers de test lus sur disque)."""

    def __contains__(self, item: object) -> bool:
        return isinstance(item, str) and bool(item)


ALL_MEDIA: frozenset[str] = _AllMedia()


def media_index(pack: Pack, studio_media_dir: Path | None = None) -> frozenset[str]:
    """Index du pack publié, complété par les médias produits par le studio pour ce pack."""
    if studio_media_dir is None:
        return pack.media_index
    studio = studio_media_dir / pack.code
    return pack.media_index | scan_media(studio) if studio.is_dir() else pack.media_index


def has_native_audio(concept: Mapping[str, Any] | None, index: Collection[str]) -> bool:
    if not concept:
        return False
    return any(
        isinstance(track, dict) and track.get("source") == "native" and track.get("src") in index
        for track in concept.get("audio") or []
    )


def has_pitch_ref(item: Mapping[str, Any] | None, index: Collection[str]) -> bool:
    """`pitchRef` d'une étape ou `pitch` d'un concept présent dans l'index."""
    if not item:
        return False
    ref = item.get("pitchRef") if item.get("pitchRef") is not None else item.get("pitch")
    return isinstance(ref, str) and ref in index


def speech_reference_available(
    step: Mapping[str, Any], concepts: Mapping[str, Mapping[str, Any]], index: Collection[str]
) -> bool:
    """Référence F0 effective d'une étape orale (comme buildExercise : `step.pitchRef ?? concept.pitch`)."""
    concept = concepts.get(str(step.get("concept", "")))
    if step.get("type") == "speak_repeat" and isinstance(step.get("pitchRef"), str):
        return step["pitchRef"] in index
    return has_pitch_ref({"pitch": concept.get("pitch")} if concept else None, index)


def step_audio_concepts(step: Mapping[str, Any]) -> list[str]:
    """Concepts dont l'audio est joué par une étape tonale (toutes les cibles possibles d'une paire minimale)."""
    if step.get("type") == "tone_minimal_pair":
        return [c for c in step.get("audioConcepts") or [] if isinstance(c, str)]
    concept = step.get("concept")
    return [concept] if isinstance(concept, str) else []


def tone_step_playable(
    step: Mapping[str, Any], concepts: Mapping[str, Mapping[str, Any]], index: Collection[str]
) -> bool:
    """Étape tonale d'écoute jouable : chaque audio qu'elle peut jouer existe en natif."""
    ids = step_audio_concepts(step)
    if step.get("type") == "tone_minimal_pair" and len(ids) != len(step.get("pair") or []):
        return False  # paire sans audio pour chaque forme : synthèse vocale, jamais notée
    return bool(ids) and all(has_native_audio(concepts.get(cid), index) for cid in ids)


def exam_item_graded(
    step: Mapping[str, Any], concepts: Mapping[str, Mapping[str, Any]], index: Collection[str]
) -> bool:
    """Item d'examen noté (§1) : oral sans référence F0 et écoute tonale sans audio natif → non noté."""
    kind = step.get("type")
    if kind in SPEECH_STEP_TYPES:
        return speech_reference_available(step, concepts, index)
    if kind in TONE_STEP_TYPES:
        return tone_step_playable(step, concepts, index)
    return True

File: packages/core/src/media.ts
import type { Concept, ContentIndex, Dialogue, GameId, Lesson, LessonStep, StepType } from "./types.ts";
import { NATIVE_AUDIO_STEP_TYPES, TONAL_STEP_TYPES } from "./types.ts";

/**
 * Disponibilité des médias (contrat phase5-parcours §1). Le bundle d'un pack expose
 * `mediaIndex` : chemins relatifs des fichiers réellement présents (`audio/…`, `pitch/…`, `img/…`).
 * Règles identiques côté API (apps/api) : toute modification se fait des deux côtés.
 */

export type MediaIndex = ReadonlySet<string>;

/** Index des médias d'un contenu ; `null` = inconnu (fichiers de test lus sur disque) → tout est considéré présent. */
export function contentMedia(content: Pick<ContentIndex, "mediaIndex">): MediaIndex | null {
  return content.mediaIndex ?? null;
}

const present = (path: string | undefined | null, media: MediaIndex | null): boolean =>
  typeof path === "string" && path.length > 0 && (media === null || media.has(path));

/** Au moins une piste `source: "native"` dont le fichier est dans l'index. */
export function hasNativeAudio(concept: Pick<Concept, "audio">, media: MediaIndex | null): boolean {
  return concept.audio.some((a) => a.source === "native" && present(a.src, media));
}

/** Courbe F0 de référence présente : `pitch` d'un concept ou `pitchRef` d'une étape. */
export function hasPitchRef(item: { pitch?: string | undefined; pitchRef?: string | undefined }, media: MediaIndex | null): boolean {
  return present(item.pitchRef ?? item.pitch, media);
}

/** Chemin de la courbe utilisée par une étape orale (étape `pitchRef`, sinon `concept.pitch`). */
export function stepPitchPath(content: ContentIndex, step: LessonStep): string | null {
  if (step.type === "speak_repeat") return step.pitchRef ?? content.concepts.get(step.concept)?.pitch ?? null;
  if (step.type === "tone_produce") return content.concepts.get(step.concept)?.pitch ?? null;
  return null;
}

/** Concepts dont l'audio porte une étape tonale (tone_minimal_pair sans `audioConcepts` : aucun, audio de synthèse). */
function tonalAudioConcepts(content: ContentIndex, step: LessonStep): Concept[] | null {
  switch (step.type) {
    case "tone_identify":
    case "tone_produce": {
      const c = content.concepts.get(step.concept);
      return c ? [c] : [];
    }
    case "tone_minimal_pair":
      return (step.audioConcepts ?? []).flatMap((id) => content.concepts.get(id) ?? []);
    default:
      return null;
  }
}

/** Une étape tonale a-t-elle son audio natif (tous ses concepts audio, au moins un) ? Étape non tonale : true. */
export function tonalStepHasNativeAudio(content: ContentIndex, step: LessonStep, media: MediaIndex | null = contentMedia(content)): boolean {
  if (!TONAL_STEP_TYPES.has(step.type)) return true;
  const concepts = tonalAudioConcepts(content, step) ?? [];
  if (step.type === "tone_minimal_pair" && (step.audioConcepts?.length ?? 0) !== step.pair.length) return false;
  return concepts.length > 0 && concepts.every((c) => hasNativeAudio(c, media));
}

/** Une étape de ce type est retirée de la séance si son enregistrement natif manque. */
export function stepRequiresNativeAudio(type: StepType): boolean {
  return NATIVE_AUDIO_STEP_TYPES.has(type);
}

/** Dialogue écoutable : au moins un tour, et chaque tour a son enregistrement présent. */
export function dialogueHasAudio(dialogue: Dialogue | undefined, media: MediaIndex | null): boolean {
  return dialogue !== undefined && dialogue.turns.length > 0 && dialogue.turns.every((t) => present(t.audio, media));
}

/**
 * L'étape a-t-elle l'audio natif qu'elle exige (contrat phase6 §1) ? Étape sans exigence : true.
 * Couvre les étapes tonales, `listen_transcribe` (on ne transcrit pas de la synthèse) et
 * `listen_gist` (le dialogue doit être enregistré en entier).
 */
export function stepHasRequiredAudio(content: ContentIndex, step: LessonStep, media: MediaIndex | null = contentMedia(content)): boolean {
  if (!stepRequiresNativeAudio(step.type)) return true;
  if (step.type === "listen_transcribe") {
    const concept = content.concepts.get(step.concept);
    return concept !== undefined && hasNativeAudio(concept, media);
  }
  if (step.type === "listen_gist") return dialogueHasAudio(content.dialogues.get(step.dialogue), media);
  return tonalStepHasNativeAudio(content, step, media);
}

/**
 * Le pack a-t-il **au moins un** enregistrement natif présent ?
 *
 * Question grossière, mais c'est la bonne : tant qu'aucune voix n'a été enregistrée, la moitié des
 * écrans affiche « Audio natif pas encore enregistré » en rouge, exercice après exercice, comme si
 * chaque mot avait un problème particulier. Ce n'est pas un incident par mot, c'est un état du
 * pack — et il se dit **une fois**, calmement.
 *
 * `media === null` (contenu lu sur disque en test) : tout est réputé présent.
 */
export function packHasNativeAudio(content: Pick<ContentIndex, "concepts" | "mediaIndex">, media: MediaIndex | null = contentMedia(content)): boolean {
  if (media === null) return true;
  for (const concept of content.concepts.values()) if (hasNativeAudio(concept, media)) return true;
  return false;
}

/**
 * Mini-jeux qui n'ont aucun sens sans voix native. `cho_noi` fait trier des barques à l'oreille par
 * leur ton : en voix de synthèse, les tons sont faux (spec §7.4), donc le jeu exige du natif et,
 * sans lui, s'ouvre sur « Pas assez de mots avec un audio natif pour jouer ici ». Mieux vaut ne pas
 * l'ouvrir : une porte qui mène à un mur n'est pas une porte.
 *
 * Les autres jeux acceptent la synthèse (le contenu n'y est pas tonal) et restent jouables.
 */
export const NATIVE_AUDIO_GAMES: ReadonlySet<GameId> = new Set<GameId>(["cho_noi", "karaoke_tonal"]);

/**
 * Le mini-jeu est-il jouable avec les médias présents ? `toneFallback` (bêta interne
 * `VITE_TTS_TONE_FALLBACK`) rend la synthèse acceptable pour les tons : `cho_noi` redevient alors
 * jouable, exactement comme les étapes tonales — c'est la même décision, prise au même endroit.
 */
export function isGamePlayable(content: ContentIndex, game: GameId, media: MediaIndex | null = contentMedia(content), toneFallback = false): boolean {
  return toneFallback || !NATIVE_AUDIO_GAMES.has(game) || packHasNativeAudio(content, media);
}

export interface PlayableOptions {
  /** Build `VITE_TTS_TONE_FALLBACK=true` (bêta interne) : étapes tonales jouées en synthèse vocale. */
  toneFallback?: boolean;
  media?: MediaIndex | null;
}

/**
 * Étape jouable en séance : une étape qui exige un enregistrement natif et ne l'a pas est retirée
 * (ni affichée ni notée), sauf repli de synthèse. Les étapes orales (`speak_repeat`, `speak_answer`,
 * `speak_roleplay`) sans courbe F0 restent jouables : elles deviennent de l'écoute non notée.
 */
export function isStepPlayable(content: ContentIndex, step: LessonStep, options: PlayableOptions = {}): boolean {
  // Le repli de synthèse ne concerne que les tons : on ne fait jamais transcrire ni écouter un
  // dialogue en voix de synthèse.
  if (options.toneFallback && TONAL_STEP_TYPES.has(step.type)) return true;
  const media = options.media === undefined ? contentMedia(content) : options.media;
  // Un mini-jeu qui s'ouvrirait sur « pas assez de mots pour jouer » est retiré comme une étape
  // d'écoute sans enregistrement : on ne fait pas traverser un cul-de-sac.
  if (step.type === "game") return isGamePlayable(content, step.game, media, options.toneFallback ?? false);
  return stepHasRequiredAudio(content, step, media);
}

/** Index des étapes jouables d'une leçon (ordre conservé, indices d'origine). */
export function playableStepIndexes(content: ContentIndex, lesson: Lesson, options: PlayableOptions = {}): number[] {
  return lesson.steps.flatMap((step, i) => (isStepPlayable(content, step, options) ? [i] : []));
}
